### ISSUE TITLE
"null" ref in onDragStart

### ISSUE THREAD

**Issue by Kuzurisisi** [clauderic/dnd-kit#936]
I need to get the width of the DOM element in the "onDragStart" event. And I pass "ref" in the "data" object. But I don't know why always on first time the "ref" is "null". Any other "onDragStart" event show correct "ref". Is this some kind of bug? Or maybe I am doing something wrong?

Sandbox: https://codesandbox.io/s/jovial-cookies-etpd8s
Variable: args.active.data.current.refOne is null on first time

**Comment by RemyMachado** [comment]
> Is this some kind of bug?

No, this not a bug, it's the way React.js works. I'll try to explain why and I'll give you a solution:

On the first render, you are passing a `null` value to `useDraggable`'s `data`, because the hook is taking the value before the `div` had a chance to set the ref.
Then, only on the second render, the hook recomputes the value you are passing him and have a truthy `ref`.

In order to get the `ref` value without awaiting for a second render, you can give the entire ref object to `useDraggable` which won't be null on the first render. Here's [how to do it](https://codesandbox.io/s/async-hooks-dfw1w2).

The important part:
```jsx
export default function Draggable(props) {
  const r = useRef(null);

  const { attributes, listeners, setNodeRef, transform } = useDraggable({
    id: "draggable",
    data: {
      refOne: r // here, pass the whole ref
    }
  });
  const style = transform
    ? {
        transform: `translate3d(${transform.x}px, ${transform.y}px, 0)`
      }
    : undefined;

  return (
    <button ref={setNodeRef} style={style} {...listeners} {...attributes}>
      <div ref={r} />
      Text
    </button>
  );
}

**Comment by Kuzry** [comment]
> 

Now I understand. The problem is that "null" is not a reference.

Thanks for the help.

**Comment by clauderic** [comment]
This issue was opened against a previous version of dnd-kit. The library has since undergone a major rewrite.

It has been rebuilt from the ground up with support for **multiple frameworks** (React, Vue, Svelte, Solid, and vanilla JavaScript), improved performance, and a new API.

- **Documentation**: https://dndkit.com
- **Migration guide**: https://dndkit.com/react/migration
- **GitHub**: https://github.com/clauderic/dnd-kit

If this issue is still relevant to the latest version, please open a new issue.
